Misc: SuperBlock, WAL, Replica, VOPR

sentientwaffle opened the pull request:
Apologies for the enormous PR; this is everything that was waiting for [`dj-vopr-workload`](https://github.com/tigerbeetledb/tigerbeetle/pull/141) to merge. Many of these changes are entangled with storage verification improvements. Others just would have had conflicts otherwise.

### Misc

- [x] CI: Run reliably-passing fuzzers.
- [x] `build.zig`:
  - [x] Clean up benchmark and fuzzer "namespaces".
  - [x] `git_commit()` is used to build the git commit hash into `tigerbeetle version`'s output at compile time.
- [x] Document `linked_event_chain_open` and `overflows_timeout`.
- [x] Add `tigerbeetle version` subcommand.
- [x] `BlockType` for grids. (This is going to be reworked in a future PR to give grid blocks a distinct "magic number" from ordinary messages.)

### Cleanup

- [x] Remove `benchmark_array_search.zig` — I think this file was duplicated somehow when LSM was merged. It is also in `src/lsm/eytzinger_benchmark.zig`.
- [x] Remove `perf.zig` — this was a backport of an `0.9` feature to `0.8`, which we don't still need.

### StateMachine

- [x] `StateMachine`: add `overflows_timeout` transfer error to catch when `timestamp + timeout` would overflow.

### Utils

- [x] `util.zig`: `copy_disjoint`, `copy_left`, `copy_right` helpers (`std.mem.copy()/copyBackwards()` but with extra assertion). Use it everywhere instead of `mem.copy{,Backwards}`.
- [x] In `NodePool.deinit()`/`MessagePool.deinit()`, assert that all nodes/messages have been restored to the pool.
- [x] `RingBuffer`: Export `count_max` declaration.

### Tests

- [ ] Fuzz the `ManifestLog`.
- [ ] Fuzz the `SuperBlock`
- [x] Unit test `SuperBlock` quorums.
- [x] Unit/fuzz test `SuperBlock` repair.

### VOPR
- [ ] `Storage` (`src/test/storage.zig`)
  - [x] Document the per-zone fault injection patterns.
  - [x] Add superblock zone fault injection. (This is used by both the superblock fuzzer and the VOPR).
  - [x] Coalesce extra arguments into `Options` with better defaults, since so many tests/fuzzers are using the `Storage` now (not just the VOPR).
  - [x] Add "crash faults": when we simulate a crash of a replica, fault the targets of in-flight writes.
  - [x] `generate_faulty_areas` is now `generate_faulty_wal_areas`. Other zones will use different fault injection patterns, based on what their repair methods can tolerate.
  - [x] Disable "crash faults" when `replica_count == 1` — we need to implement redundant WAL header repair for that to work (it is an existing TODO).
  - [x] Add `Storage.copy()`, which is used by the superblock and manifestlog fuzzers to clone/"fork" a `Storage` so they can try to open + repair from it, before we use the old storage to continue on our way.
  - [x] Add back Jamie's `size_used()` function. (I can't find the commit it was in, but hopefully this is right).
  - [ ] Add `verify_write_wal_header`: when a redundant header is written, verify that we didn't leak it before the prepare.
  - [ ] Add `verify_write_superblock`: when a superblock sector is written, verify its trailers are already on-disk. Also checks that superblock writes are properly serialized (to guard against correlated errors).
  - [x] Add some helper functions for reading from the storage, e.g. `storage.grid_block(address)`. These can be used by `StorageChecker` and fuzzers.
- [ ] `MessagePool`/`Network`/`PacketSimulator`: Use a separate message pool for the network so that we don't have to allocate an extra 20 messages in each replica/client pool. (As an unexpected bonus, this dramatically simplifies the message-leak check in `Cluster.crash_replica()`.)
- [ ] `StorageChecker`: This works, but too well :wink:. It is disabled until compaction's concurrent block allocation issue is fixed, so that the VOPR can still be useful and not crash right away.

### WAL

- [x] Documentation!
- [x] Don't leak redundant headers before a prepare is on disk.
- [x] Remove the [TODO](https://github.com/tigerbeetledb/tigerbeetle/commit/f30d23c8312478663db9afd777dc245c00810b43): `When participating in a view change, each replica should only send the headers from its working set that it knows it prepared.` — this turns out to be safe as-is.
- [x] Split the `Zone.wal` into `Zone.wal_headers` and `Zone.wal_prepares`. This enables more fine-grained fault injection and verification in the `Storage`.
- [x] Change recover case `@D` when `replica_count=1` from `vsr` to `fix`.
- [x] When we know the size of the message we are reading, use a smaller read buffer.
- [x] [Improve torn repair recovery](https://github.com/tigerbeetledb/tigerbeetle/commit/1b791ef115775a26bcf64ab81efb798ab2a0ce75)
  - [ ] This also checks the "only one concurrent write to the WAL" invariant for `replica_count=1`.
  - [ ] Add a TODO: For very small WALs (not usually a concern) we must be careful not to write to all redundant header sectors simultaneously, since a crash could leave us without redundant headers (which is very bad for a single-replica cluster).
- [x] [Change case `@D` for replica_count=1 from `vsr` to `fix`](https://github.com/tigerbeetledb/tigerbeetle/commit/04941070d7da39d83c7432a700b45c275b05859f) to allow a torn header to be recovered.
- [x] Add `Journal.remove_entry()` — This is used in one (tricky) case in `repair_prepares` to discard an op that we don't need and can't repair.
- [x] Make `Journal.writing()` [more conservative](https://github.com/tigerbeetledb/tigerbeetle/commit/e9478db4f773622a12aa036775925e5e795e7f83).

### Replica

- [x] Document additional invariants.
- [x] Add `on_compact()` and `on_checkpoint()` handlers (like `on_change_state`); used by the `StorageChecker`.
- [x] Use `SuperBlock.vsr_state.commit_min_checksum` to anchor the hash chain — since `vsr_state.commit_min` is checkpointed, we don't want to rely on it being intact in our WAL in order to commit, and we just need the checksum to hash-chain the next commit.
  - [x] Use this in `valid_hash_chain_between`.
- [x] Additional assertions on outgoing messages.

### SuperBlock

- [x] Document `VSRState` invariants.
- [x] Remove `Magic`, didn't do anything.
- [x] Add `SuperBlock.vsr_state.commit_min_checksum` (see Replica).
- [x] Remove the `SuperBlock.staging` sector — this isn't needed since both `view_change()` and `checkpoint()` now store their update (`vsr_state`) on the `Context`, which is queued.
- [x] Remove the extra copyset; only 4 copies now.
  - `view_change()` no longer bumps the sequence twice.
  - Only check the parent hashchain opportunistically.
- [x] Repair after `open`. See: `RepairIterator` in `superblock_quorums.zig`.
- [x] Move `Quorums` into its own file.

---

### `tigerbeetle version`

Sample non-verbose output (the "experimental" is because we aren't tagging versions yet, but eventually that would be semver):

```
TigerBeetle version experimental
```

Sample _verbose_ output:

```
TigerBeetle version experimental

git_commit="f358de668dcccc366ec6458717054d77c2f230e4"

zig_version=0.9.1
mode=Mode.Debug
os=Os{ .tag = Tag.linux, .version_range = VersionRange@21e674 }

deployment_environment=Environment.development
log_level=2
replicas_max=6
clients_max=32
quorum_replication_max=3
port=3001
address="127.0.0.1"
memory_size_max_default=1073741824
cache_accounts_max=8192
cache_transfers_max=65536
cache_transfers_pending_max=65536
journal_slot_count=1024
journal_size_max=1073872896
connections_max=38
message_size_max=1048576
message_body_size_max=1048448
pipeline_max=32
connection_delay_min_ms=50
connection_delay_max_ms=1000
connection_send_queue_max_replica=4
connection_send_queue_max_client=2
client_request_queue_max=32
tcp_backlog=64
tcp_rcvbuf=4194304
tcp_sndbuf_replica=4194304
tcp_sndbuf_client=2097152
tcp_keepalive=true
tcp_keepidle=5
tcp_keepintvl=4
tcp_keepcnt=3
tcp_user_timeout_ms=17000
tcp_nodelay=true
cache_line_size=64
sector_size=4096
direct_io=true
io_depth_read=8
io_depth_write=8
superblock_copies=4
size_max=17592186044416
block_size=65536
block_count_max=268435456
lsm_trees=30
lsm_levels=7
lsm_growth_factor=8
lsm_key_size_max=32
lsm_table_size_max=67108864
lsm_manifest_node_size=16384
lsm_batch_multiple=4
lsm_snapshots_max=32
lsm_value_to_key_layout_ratio_min=16
tick_ms=10
rtt_ticks=30
rtt_multiple=2
backoff_min_ticks=10
backoff_max_ticks=1000
clock_offset_tolerance_max_ms=10000
clock_epoch_max_ms=60000
clock_synchronization_window_min_ms=2000
clock_synchronization_window_max_ms=20000
verify=true
journal_size_headers=131072
journal_size_prepares=1073741824
is_32_bit=false
```

sentientwaffle committed:
Util: copy_left/copy_right/copy_disjoint

Add `mem.copy` helper functions with extra safety checks.

sentientwaffle committed:
WAL: Read smaller buffers when size is available

If a message's exact size is known, don't try to read the whole thing.

sentientwaffle committed:
CI: Run VOPR canary in CI

Also don't run CI twice for every PR commit.

sentientwaffle committed:
MessagePool/NodePool: On deinit verify cleanup

`MessagePool.deinit`/`NodePool.deinit` now verify that all messages/nodes were previously `unref`'d/`release`'d.
This can help detect a message/node leak.

sentientwaffle committed:
Build: Clean up and fix benchmarks

- Build benchmarks in CI, just to make sure they can compile and don't get out of date.
- Update benchmarks for Zig 0.9.
- Remove `benchmark_array_search.zig`, which was identical to `eytzinger_benchmark.zig`. I think this snuck in via a rebase/merge.
- Remove some `build.zig` duplication in anticipation of future benchmarks.
- Remove `perf.zig` (Linux perf API). This is part of stdlib now that we use Zig 0.9.
- Add minor clarification to `cli.zig` help text.

sentientwaffle committed:
CLI: "tigerbeetle version" command

sentientwaffle committed:
VSR: Stricter validation of outgoing messages

sentientwaffle committed:
WAL: Don't leak redundant headers.

Don't leak redundant headers whose corresponding prepare is not yet written.

This is incomplete: the replica still needs to refrain from sending headers (e.g. in a DVC) before the prepare is written.
That is a more involved change, to be implemented separately.

sentientwaffle committed:
WAL: Split WAL into two Zones.

Split the WAL into two `vsr.Zone`s (`wal_headers` and `wal_prepares`), rather than having `Journal` manage the two itself.

This is setup for:
- Injecting faults into the two zones independently via the test `Storage`. (The redundant headers have a much higher tolerance for corruption than the prepares).
- Verifying reads/writes to the zones via the test `Storage`. (For example, as of the last commit, we don't leak redundant headers before the prepare is written. This change will enable Storage to verify that).

sentientwaffle committed:
WAL: Verify prepare write before redundant header

sentientwaffle committed:
WAL: Safer Journal.writing()

`writing(5, checksum₁)` should return false when:

  queue[0] = { 5, checksum₁ }
  queue[1] = { 5, checksum₂ }

or when

  queue[0] = { 5,              checksum₁ }
  queue[1] = { 5 + slot_count, checksum₂ }

That is, we *are* writing the message, but that write is already due to be overwritten.

sentientwaffle committed:
WAL/VSR: Don't repair committed op without header

Don't try to repair a committed op if we don't have the header.

This is a tricky case because `find_latest_headers_break_between` is built for uncommitted headers, and we may not even have an "anchor" op on the bottom end to help repair headers. This case can only happen when the faulty message was already checkpointed (we crashed and recovered).

If we wanted to support this, another option would be to extend `request_prepare` to support requesting by op+parent (rather than op+checksum).

Other fixes:
- Fix `remove_entries()` — update the `headers_redundant` (with the new reserved header).
- Fix `TableIterator.deinit()` panic. The VOPR will call `deinit()` at any time — even mid-IO.

Remove the TODO from journal — it isn't necessary.
Documented in https://github.com/tigerbeetledb/tigerbeetle/pull/102/commits/041457fc3c8629ed612f305f42e3105e8997208e:

    /// Perhaps unintuitively, it is safe to advertise a header before its message is prepared
    /// (e.g. the write is still queued). The header is either:
    ///
    /// * committed — so another replica in the quorum must have a copy, according to the quorum
    ///   intersection property. Or,
    /// * uncommitted — if the header is chosen, but cannot be recovered from any replica, then
    ///   it will be discarded by the nack protocol.

sentientwaffle committed:
SuperBlock: Repair bad copies after open()

At the end of `open()`, repair any broken copies by writing their sector and all trailers.
This is a coarse implementation of SuperBlock repair, to fix a particular liveness issue (described below).

As an example of the liveness issue, consider the following case:

Initially the superblock sectors are:

    copy 0/8 sequence=2
    copy 1/8 sequence=2
    copy 2/8 sequence=2
    copy 3/8 sequence=2
    copy 4/8 sequence=1
    copy 5/8 sequence=1
    copy 6/8 sequence=1
    copy 7/8 sequence=1

1. Checkpoint:
  1. Write copy 4/8 with sequence=3.
  2. Write copy 5/8 with sequence=3.
  3. Crash.
2. Recover. Open superblock. Quorums are:
  - sequence=2 count=4 (threshold=2, valid)
  - sequence=1 count=2 (threshold=2, valid)
  - sequence=3 count=2 (threshold=2, valid)
3. Checkpoint: Write copies 0…3 sequence=4.
4. Checkpoint: Verify quorums:
  - sequence=4 count=4 (threshold=3, valid)
  - sequence=1 count=2 (threshold=3, invalid)
  - sequence=3 count=2 (threshold=3, invalid!)
5. Parent quorum lost.

sentientwaffle committed:
SuperBlock/LSM: Don't repeat checkpoints

- Don't checkpoint at the same op twice (e.g. after crash+recover).
- When we skip replaying compactions (in Tree) we must still compact the mutable → immutable table at the last tick.
- Document `VSRState` invariants.
- Fix `view_change()`:
  - Don't permit it to update `commit_min`.
  - In `view_change()`, check `staging.vsr_state` instead of `working.vsr_state` — the former is more up-to-date (it may have been updated by an in-progress `checkpoint()`).

sentientwaffle committed:
VSR: Wait for one message after WAL before commit

Wait for all replicas to `prepare_ok` the message _after_ the WAL wraps (in addition the one last one of the wrap) so that we know they checkpointed.
This fixes a case where they would ack but not checkpoint, then corruption prevents them from checkpointing on recover, but the other replicas have already wrapped.

sentientwaffle committed:
VOPR: Fix chance_f64's probability

sentientwaffle committed:
SuperBlock: Unit test superblock's Quorums.working

Additionally:
- Update documentation (`linked_event_chain_open`, `MessageBus` messages are always available).
- In `StateChecker`, we don't (or soon won't) require `commit_min` to exist in the WAL.
- Change the superblock's log.err()s to log.warn()s because Zig's test framework turns log.err()s into failed tests, which inhibits testing errors. Is isn't currently possible to disable this behavior.
- Add `id` assertions to `prefetch_enqueue()`.

sentientwaffle committed:
SuperBlock: Add VSRState.commit_min_checksum (WIP)

Add `VSRState.commit_min_checksum` to the superblock sector. (But it isn't used yet by the replica.)

Motivation: The working superblock's `vsr_state.commit_min` corresponds to the replica's current `op_checkpoint`. This op is always already committed (even after recovering). And it is allowed to be overwritten by the new WAL wrap — since it is committed, we don't need the message body anymore.
However, the op immediately following `op_checkpoint`/`vsr_state.commit_min` may need to be committed after a crash+recover. To commit, we need the parent's checksum so that we can verify the hash chain.

sentientwaffle committed:
Test/Storage: SuperBlock verification and faults

- Implement fault injection for the SuperBlock zone.
- On each superblock sector write, verify superblock trailers (`verify_write_superblock`).
- Add a separate `crash_fault_probability` for `Storage.reset()`. This will typically be much higher than the read/write fault probability.
- Clarify "faulty area" wrt the WAL.
- Simplify `Storage.init()`. `test/Storage` is being used by more tests (and many more coming soon). But parts of its interface are applicable only to the VOPR.
- Expose superblock's format functions, since storage needs the offsets.

sentientwaffle committed:
VOPR: Add checker to verify deterministic storage

Add StorageChecker to verify at checkpoint that storage is deterministic across replicas.

sentientwaffle committed:
SuperBlock: Fuzz open/checkpoint/view_change

Includes two fixes. Both are (now commented-out) assertions in the superblock.
The caller can modify `superblock.staging.vsr_state` any time a checkpoint is not in progress — `view_change()` should ignore it.
These are only commented out for now (rather than removed) because I may use `context.vsr_state` to solve this instead.

sentientwaffle committed:
LSM/Fuzz: Fuzz forest storage latency

sentientwaffle committed:
SuperBlock/Fuzz: Verify view_change_in_progress()

Also don't acquire a block every time.

sentientwaffle committed:
SuperBlock: Quorums.Threshold; Context checking

- Add `Quorums.Threshold` enum. Previously the threshold could be derived entirely from the caller (`open`, `checkpoint`, …). But because `open()` can write, it uses both quorum sizes depending on the situation.
- Replace undefined fields on the `Context` with checked optionals to avoid possible footguns.
- Diagram the relationship for working/staging/writing.
- Rename "Format" to "Layout" to avoid overloading the word "format" (in the superblock context).

sentientwaffle committed:
WAL: Allow recovery of torn header when replicas=1

The addition of storage faults on crash of `replica_count=1` (to simulate torn writes) exposed this improvement in the recovery table.

sentientwaffle committed:
WAL: Improve torn repair recovery

Allow a faulty `op_checkpoint` to be truncated (via `recover_torn_prepare()`) when it is safe to do so.

Additionally, for `replica_count=1`, enforce the "only one concurrent write to the WAL" constraint.

sentientwaffle committed:
SuperBlock: commit_min_checksum, checkpoint(vsr_state)

- Integrate `commit_min_checksum`, to be used as the hash chain anchor when the `commit_min` is overwritten/faulty in the WAL.
- Pass `vsr_state` into `superblock.checkpoint()` rather than updating `staging` directly. This allows it to enforce invariants more immediately. And it removes footguns of multiple/forgotten updates to `staging`.

sentientwaffle committed:
VOPR: StorageChecker

Test for storage determinism violations on checkpoint/compact. Currently disabled until the compaction determinism issue is fixed.

sentientwaffle committed:
SuperBlock: Remove magic number

Other sectors use a different schema anyway.

Use the extra space for `version`.

sentientwaffle committed:
LSM: Fuzz ManifestLog

sentientwaffle committed:
StateMachine: Guard against timeout overflow

Post/void transfer computes `pending_transfer.timestamp + pending_transfer.timeout` — which can currently overflow if the timeout is really (really) huge.
Verify that it doesn't overflow earlier, when the pending transfer is created.

This is a breaking client change — it renumbers the `CreateTransferResult`s.

sentientwaffle committed:
SuperBlock: Remove secondary copyset

- Motivation:
  - Original motivation is that having a single set of trailers will allow the new (unimplemented) client table to be updated in-place.
  - The extra superblock sector copyset was deemed unnecessary.
- Still check the parent hashchain if the parent copy is present. Usually it won't be.

sentientwaffle committed:
SuperBlock: Remove "staging" superblock sector

Since all updates come through `context.vsr_state` via `checksum()`/`view_change()`'s `vsr_state` argument, `staging` is unnecessary.

Also coalesce `write_staging` with `write_view_change`.

I kept the name `write_staging` for a lack of a better term for "the superblock we are about to write, but haven't started yet, and it isn't the working superblock". It is accurate, it just doesn't correspond to the `staging` field anymore.

sentientwaffle committed:
VOPR: Use a separate message pool for Network

sentientwaffle committed:
SuperBlock: Choose repair order carefully

Also handle corruption of the superblock sector's copy index.

sentientwaffle committed:
SuperBlock: Extract Quorums to its own file

sentientwaffle committed:
CI: Run reliably-passing fuzzers in CI

sentientwaffle committed:
Misc: Rebase

And fix the superblock quorums test — can't shuffle when using a high-copy-index.

sentientwaffle committed:
VSR: Remove double-checkpoint check

This was added as part of <https://github.com/tigerbeetledb/tigerbeetle/commit/6ec0a66feb49950d042dff3c5ec7115e8c286e04>, but I can't figure out why it is/was needed.

sentientwaffle commented:
Known problems (I'll address tomorrow):

- ~Message leak in my `PacketSimulator`/`Network` rework.~
- ~CI trouble with the `ManifestLog` workflow.~
- ~CI trouble with the `Simulator Canary` workflow (I'm going to leave it in even though it often fails right now; I'll just remove the `test` workflow's dependency on it).~

sentientwaffle committed:
Test: Fix CI; update demos

sentientwaffle committed:
MessageBus: Implement deinit()

Also fix `Client`'s `deinit()` to clean up the request queue.

sentientwaffle committed:
Simulator: Fix network queue capacity

jamii commented on .github/workflows/test_and_build.yml:
```diff
@@ -48,7 +48,54 @@ jobs:
+    steps:
+      - uses: actions/checkout@v2
+      - run: ./scripts/install_zig.sh
+      - run: zig/zig build fuzz_lsm_manifest_log
```
How many time is this run in CI? Do we want to run it with a fixed seed?

sentientwaffle replied:
Just once. It is sort of a canary, i.e. make sure that the fuzzer builds and succeeds a test.
The intent isn't to use GitHub as our fuzzer infrastructure, it is just to make sure there are no low-hanging bugs introduced by a PR.

Re: fixed seed — I was thinking that fuzzer errors will (hopefully!) be rare enough that it wouldn't matter, and varying the seed can give us a tiny bit extra coverage, but I'm open to changing it.

jamii replied:
I guess let's leave it random until it's a problem.

sentientwaffle replied:
Thinking about this more I think you're right, we should use a fixed seed so that we never have the PR pass CI, then fail it only after it is merged.

jamii commented on build.zig:
```diff
@@ -4,13 +4,24 @@ const builtin = @import("builtin");
+        } else {
+            options.addOption(?[]const u8, "git_commit", null);
+        }
+        tigerbeetle.addOptions("tigerbeetle_build_options", options);
```
Woah, I did not know about this.

sentientwaffle replied:
:laughing: It isn't documented anywhere I could find. I'm thinking this is also how we will support comptime configuration from the user (rather than what we do right now, which is just manually modifying `config.zig`).

Zig has a couple of open issues that will make this even easier from the source code side: https://github.com/ziglang/zig/issues/1646 and https://github.com/ziglang/zig/issues/226.

sentientwaffle committed:
Simulator: Fix integer overflow

Also use a fixed seed for CI so that we can never have a passing PR that fails on `main` when it is merged.

jamii commented on src/test/storage.zig:
```diff
@@ -149,41 +179,93 @@ pub const Storage = struct {
+        const sector_max = sector_iterator.next() orelse 0;
+        return (sector_max + 1) * config.sector_size;
+    }
+
```
I was using memory_written.count() instead. I guess this calculation is assuming that unwritten sectors before the last last written sector still occupy storage, which leads me to realize that I don't know how file-systems work.  If I open a file and write to sector 1000000, do the preceding 999999 sectors get created on disk?

sentientwaffle replied:
That's a good point, we use `fallocate` and I'm not sure how it works at that level :thinking:.

jorangreef replied:
It depends on the file system. I think most of the time, the file would be sparse, and only the last sector would be written to disk. i.e. The underlying sectors referenced by a file are not guaranteed to be contiguous. How this impacts the filesystems view of available storage on disk might be a separate matter.

OSTEP is great (along with any storage paper from the Arpaci-Dusseau's): https://pages.cs.wisc.edu/~remzi/OSTEP/

sentientwaffle replied:
Perfect, I'll change this to `.count()` then. Thanks @jamii & @jorangreef!
(Also, how did I not know that OSTEP was written by the Arpaci-Dusseau's?!)

jorangreef replied:
Thanks @sentientwaffle for the ping!

Re: `fallocate()` that would allocate the sectors, and ensure that they are zeroed. But there are optimizations in how it does this: https://man7.org/linux/man-pages/man2/fallocate.2.html

jamii commented on src/ewah.zig:
```diff
@@ -2,7 +2,9 @@ const std = @import("std");
 const assert = std.debug.assert;
 const math = std.math;
 const mem = std.mem;
-const div_ceil = @import("util.zig").div_ceil;
```
Could util.div_ceil be replaced with `std.math.divCeil(...) catch unreachable`?

sentientwaffle replied:
See discussion here: https://github.com/tigerbeetledb/tigerbeetle/pull/62#discussion_r791035331

jamii replied:
I see :)

jamii commented on src/lsm/table_iterator.zig:
```diff
@@ -90,8 +91,6 @@ pub fn TableIteratorType(comptime Table: type, comptime Storage: type) type {
         }
 
         pub fn deinit(it: *TableIterator, allocator: mem.Allocator) void {
-            assert(!it.read_pending);
```
Was this not necessary? (Or maybe it makes it hard to simulate a crash and restart if deinit panics?)

sentientwaffle replied:
> (Or maybe it makes it hard to simulate a crash and restart if deinit panics?)

That's it exactly. We deinit the replica (and all the way down) when simulating a crash — which can occur at any time.

jamii replied:
Is it possible that that would result in the io layer delivering a read to a freed table?

sentientwaffle replied:
For these purposes, we would be using the test storage (`src/test/storage.zig`) which is running in our same thread. When we "crash" a replica, we call `Storage.reset()`, which clears the pending reads/writes (and as of this PR, injects faults into some of the write targets). If it did let the write finish that would be ok too, though (as long as it didn't call the write's callback).

jamii replied:
For crashes, would it be easier to just deinit the arena allocator? After all, we won't call deinit in real crashes either.

sentientwaffle replied:
We don't actually use an arena allocator within the VOPR. But even if we did, having this extra granularity is useful:

- e.g. it lets us reuse the message pool instead of freeing/allocating it, which saves some time. (The message pool is a whole bunch of 1MB allocations).
- we can test `deinit()` functions — even though the real replica doesn't ever use it, it is still nice to have.

sentientwaffle replied:
The crash logic is here by the way: https://github.com/tigerbeetledb/tigerbeetle/blob/55cdb208814aa9ac49ec7521c428653eda347392/src/test/cluster.zig#L281-L321

It is simplified a little in this PR (around counting packets in the network). Most of the complexity is around determining whether or not we can safely crash a replica, to ensure that VSR recovery protocol will be able to bring it back.

It will be dramatically simplified within the next month or so when we switch to using the superblock for recovery instead.

jamii commented on src/lsm/tree_fuzz.zig:
```diff
@@ -339,19 +349,10 @@ pub fn run_fuzz_ops(fuzz_ops: []const FuzzOp) !void {
             .read_latency_min = 0,
             .read_latency_mean = 0,
             .write_latency_min = 0,
             .write_latency_mean = 0,
```
In forest_fuzz you set *_latency_mean. Do you want to do the same here? Or is there no way for reads/writes to interleave here?

sentientwaffle replied:
Good point; I'll update it. When there are multiple compactions (i.e. from different levels) running in parallel, then IOs will interleave.

jamii commented on src/message_bus.zig:
```diff
@@ -145,8 +146,12 @@ fn MessageBusType(comptime process_type: vsr.ProcessType) type {
-        /// TODO This is required by the Client.
-        pub fn deinit(_: *Self, _: std.mem.Allocator) void {}
+        pub fn deinit(bus: *Self, _: std.mem.Allocator) void {
+            for (bus.connections) |*connection| {
```
A bunch of stuff is allocated in init. Should it be deallocated here?

jamii commented on src/state_machine.zig:
```diff
@@ -1837,7 +1834,7 @@ test "create/lookup/rollback transfers" {
                 .credit_account_id = 200,
-                .timeout = 1,
+                .timeout = (std.math.maxInt(u64) - timestamp) + 1,
                 .ledger = 0,
```
Are these intended to trigger overflows_timeout? Or just test that it's the last thing checked?

sentientwaffle replied:
The idea with these tests is that they are checking the error code precedence, at the same time as they check the different error cases. In this particular case, we are checking that a `timeout_reserved_for_pending_transfer` error takes precedence over the `overflows_timeout` error (i.e. the former is the result) even though both are actually applicable to this transfer.

jamii replied:
Oh, oops, I see the comment way up at the top :)

jamii commented on src/test/storage.zig:
```diff
@@ -1,10 +1,35 @@
+//!
+//! Fault Injection
+//!
+//! Storage injects faults that the cluster can (i.e. should be able to) recover from.
```
I suppose eventually we should also have a separate set of tests for storage faults beyond these limits, to ensure that the cluster detects them and shuts down.

sentientwaffle replied:
Often we don't actually want to shut down — WAL repair involves trying to get the message from other replicas. And storage faults can be transient (e.g. latent sector errors). So retrying _may_ eventually succeed. But you're right, we should test that more thoroughly.

jamii commented on src/test/storage.zig:
```diff
@@ -33,12 +61,23 @@ pub const Storage = struct {
-        write_fault_probability: u8,
+        /// Chance out of 100 that a read will corrupt a sector, if the target memory is within
+        /// a faulty area of this replica.
+        read_fault_probability: u8 = 0,
```
Elsewhere we use floats for probabilities (eg cluster.HealthOptions). Why the u8 here?

sentientwaffle replied:
Because I didn't realize that when I wrote `HealthOptions`. :wink: We want to eventually migrate away from using any floats at all though — they aren't deterministic between different CPUs, and we want all of our simulators/fuzzers to be fully deterministic.

jamii replied:
Hmmm. Maybe I should avoid them in test/fuzz too.

sentientwaffle committed:
MessageBus: fix deinit

Also:
- Add `BuildMode` and Zig's version to `tigerbeetle version`.
- Add latency & seed to fuzzer test storage.
- Change `Storage.size_used` to treat storage as sparse.

jamii commented on src/test/storage.zig:
```diff
@@ -456,33 +544,280 @@ pub const Storage = struct {
-        // The read/write does not touch any faulty sectors.
-        if (start >= end) return null;
+            const offset_faulty =
+                faulty_wal_areas.first_offset + (offset_in_storage / period) * period;
```
```suggestion
                faulty_wal_areas.first_offset + @divTrunc(offset_in_storage, period) * period;
```

A little more obvious.

jamii replied:
I kind of hope that zig removes integer division via `/` entirely in favor of explicit use of `@divTrunc` etc.

sentientwaffle replied:
I try to avoid `/` when I remember, but it isn't automatic for me yet.

jamii commented on src/vsr/superblock.zig:
```diff
@@ -108,19 +114,35 @@ pub const SuperBlockSector = extern struct {
+            };
+        }
+
         pub fn internally_consistent(state: VSRState) bool {
```
`internally_consistent` and `monotonic` are only ever called inside asserts. Would it make more sense for them to be `verify_internally_consistent` and `verify_monotonic` and assert internally? As a bonus, we'd then get to see exactly which assert failed.

sentientwaffle replied:
We do use `monotonic()` as a boolean in `superblock_quorums.zig`, when we are trying to pick the working superblock quorum.

Also, one benefit of returning `bool` is it is more amenable to testing (at least until Zig supports testing panics). (We don't have these here, but we probably should since they are so critical for correctness.)

jamii commented on src/vsr/superblock_quorums.zig:
```diff
@@ -0,0 +1,625 @@
+    result: QuorumsType(.{ .superblock_copies = 4 }).Error!u64,
+) !void {
+    const Quorums = QuorumsType(.{ .superblock_copies = 4 });
+    var prng = std.rand.DefaultPrng.init(@intCast(u64, std.time.milliTimestamp()));
```
Do you want to split this like segmented_array_fuzz? Fixed seed for unit tests and a `main` that tries other seeds.

jamii replied:
Similarly below.

sentientwaffle committed:
SuperBlock: Move Quorum fuzzes to separate file

This way unit tests are deterministic.

jorangreef commented on src/util.zig:
```diff
@@ -49,3 +38,100 @@ test "div_ceil" {
+pub inline fn copy_left(
+    comptime precision: CopyPrecision,
+    comptime T: type,
+    destination: []T,
```
```suggestion
    target: []T,
```

Convention that we use elsewhere, lines up with `source`.

jorangreef commented on src/util.zig:
```diff
@@ -49,3 +38,100 @@ test "div_ceil" {
+    try std.testing.expect(std.mem.eql(usize, a, &.{ 2, 3, 4, 5, 6, 7, 6, 7 }));
+}
+
+pub inline fn copy_right(
```
I like that you've gone with inline here.

jorangreef commented on src/util.zig:
```diff
@@ -49,3 +38,100 @@ test "div_ceil" {
+        .inexact => assert(destination.len >= source.len),
+    }
+
+    if (!is_disjoint(T, T, destination, source)) {
```
I think we should move from the `is_` boolean prefix, since these functions are typically used in conditions, and Zig is different to most languages here in that conditions are always clearly booleans.

So it reads nicely, as one would pronounce it: `if (disjoint_slices())`

There was a post a few years back, perhaps by Steve Yegge I can't recall, but it was on something similar for Java. At the time the convention was getters/setters, so `get_value()` and `set_value()` and the insight was that the read path could be just `value()` since the write path is otherwise clearly indicated. So `value()` and `set_value()`.

jorangreef commented on src/vsr/journal.zig:
```diff
@@ -7,29 +7,57 @@ const math = std.math;
 
-/// There are two contiguous circular buffers on disk in the journal storage zone (`vsr.Zone.wal`).
+/// The WAL consists of two contiguous circular buffers on disk: `vsr.Zone.wal_headers` and
+/// `vsr.Zone.wal_prepares`.
```
```suggestion
/// The WAL consists of two contiguous circular buffers on disk:
/// - `vsr.Zone.wal_headers`
/// - `vsr.Zone.wal_prepares`
```

Since we are wrapping.

jorangreef commented on src/vsr/journal.zig:
```diff
@@ -7,29 +7,57 @@ const math = std.math;
+        switch (ring) {
+            .headers => {
+                comptime assert(config.sector_size % @sizeOf(Header) == 0);
+                const offset_in_ring = vsr.sector_floor(slot.index * @sizeOf(Header));
```
```suggestion
                const ring_offset = vsr.sector_floor(slot.index * @sizeOf(Header));
```

jorangreef replied:
So that the qualifier is last.

jorangreef commented on src/vsr/journal.zig:
```diff
@@ -572,11 +612,12 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
         ) ?HeaderRange {
-            assert(op_min <= op_max);
+            assert(op_max >= op_min);
+            assert(op_max - op_min < slot_count);
```
Since `op_max/min` are both inclusive indexes, I think we need to convert to a count by adding 1?

i.e. If `op_min=2` and `op_max=3` then `3-2=1` when in fact the count of ops involved is 2?

And then use `<=` in the comparison, since we are now comparing like for like, counts on both sides. It's the same effect but it seems better to keep the comparison in terms of the same units.

sentientwaffle replied:
Updated to:

```zig
            assert(op_max - op_min + 1 <= slot_count);
```

jorangreef commented on src/vsr/journal.zig:
```diff
@@ -732,6 +775,9 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
+                    message_size = vsr.sector_ceil(exact.size);
+                    assert(message_size <= config.message_size_max);
                 }
             }
```
Do we need to set `message_size` in the else branch here too?

sentientwaffle replied:
No, the `message_size` here is just an optimization — if we know the exact size of the message, we can read fewer than `message_size_max` bytes. If we don't know the size, we just read the full `message_size_max` bytes.

(I'll add a comment to explain this in the code too).

jorangreef commented on src/vsr/journal.zig:
```diff
@@ -751,20 +797,18 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
                 buffer,
-                .wal,
-                offset,
+                .wal_prepares,
+                Ring.prepares.offset(slot),
```
Nice!

jorangreef commented on src/vsr/journal.zig:
```diff
@@ -1108,7 +1153,7 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
         ///   match view               _   _   _   _   _   _   _   _   _  !1   _   _  !0  !1
         ///   decision (replicas>1)  vsr vsr vsr vsr vsr fix vsr fix vsr nil fix vsr vsr eql
-        ///   decision (replicas=1)                  fix     fix
+        ///   decision (replicas=1)              fix fix     fix
```
Win!

jorangreef commented on src/vsr/journal.zig:
```diff
@@ -1237,32 +1289,30 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
             // We must be certain that the torn prepare really was being appended to the WAL.
-            // Return if any faults do not lie between the checkpoint and the torn prepare, such as:
+            // Return null if any faults do not lie between the checkpoint and the torn prepare,
+            // such as:
```
Would be great just to explain "why" a little more here for a first time reader.

i.e. "and the torn prepare, such as: ... Since then we cannot be certain if the prepare was torn or corrupted.

jorangreef commented on src/vsr/journal.zig:
```diff
@@ -1308,20 +1358,16 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
                         self.dirty.clear(slot);
                         // TODO Repair header on disk to restore durability.
                     } else {
+                        assert(prepare.?.command == .prepare);
+                        assert(self.prepare_inhabited[slot.index]);
+                        assert(self.prepare_checksums[slot.index] == prepare.?.checksum);
```
Do these not apply to the first branch any longer?

sentientwaffle replied:
Correct, case `@D` will `fix` a reserved header.

jorangreef commented on src/vsr/journal.zig:
```diff
@@ -1415,28 +1460,35 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
+            const replica = @fieldParentPtr(Replica, "journal", self);
+
+            self.headers[slot.index] = Header.reserved(replica.cluster, slot.index);
+            self.headers_redundant[slot.index] = self.headers[slot.index];
```
We could make these lines safer w.r.t. line ordering by caching `reserved = Header.reserved` and then using it for both.

jorangreef commented on src/vsr/journal.zig:
```diff
@@ -1460,6 +1512,8 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
 
         /// `write_prepare` uses `write_sectors` to prevent concurrent disk writes.
+        // TODO To guard against torn writes, don't write simultaneously to all redundant header
+        // sectors. (This is mostly a risk for single-replica clusters very small WALs).
```
```suggestion
        // sectors. (This is mostly a risk for single-replica clusters with small WALs).
```

jorangreef replied:
I like how you say "single-replica clusters". That's exactly it. Everything is a batch or a cluster.

jorangreef commented on src/vsr/journal.zig:
```diff
@@ -1485,6 +1540,7 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
                 assert(!self.faulty.bit(slot));
                 assert(self.prepare_inhabited[slot.index]);
                 assert(self.prepare_checksums[slot.index] == message.header.checksum);
+                assert(self.headers_redundant[slot.index].checksum == message.header.checksum);
```
Now we're really lining up! Awesome.

jorangreef commented on src/vsr/journal.zig:
```diff
@@ -1537,6 +1592,7 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
                 assert(!self.prepare_inhabited[slot.index]);
                 self.prepare_inhabited[slot.index] = true;
                 self.prepare_checksums[slot.index] = message.header.checksum;
+                self.headers_redundant[slot.index] = message.header.*;
```
Have we audited all the places where we assign to `prepare_checksums` to check that we do the same for `headers_redundant`? Is there way we can enforce this through a helper?

sentientwaffle replied:
:thinking: I don't think we should enforce that — they aren't necessarily updated in sync.

- `prepare_checksums[slot]` is the checksum of the prepare that is in the prepares ring — even if it isn't in `journal.headers` (e.g. because it was truncated by a view change). This maximizes the replica's ability to assist other replicas with repair.
- `headers_redundant[slot]` is the checksum of the prepare in the prepares ring _only when_ it is in our `journal.headers` as well.

Looking at this code I think the `assert(!self.prepare_inhabited[slot.index]);` isn't actually correct in the "write A→B→A" case. I'm rewriting it to:

```zig
            {
                // `prepare_inhabited[slot.index]` is usually false here, but may be true if two
                // (or more) writes to the same slot were queued concurrently and this is not the
                // first to finish writing its prepare.
                const slot = self.slot_for_header(message.header);
                self.prepare_inhabited[slot.index] = true;
                self.prepare_checksums[slot.index] = message.header.checksum;
            }

            if (self.slot_with_op_and_checksum(message.header.op, message.header.checksum)) |slot| {
                self.headers_redundant[slot.index] = message.header.*;
            } else {
                self.write_prepare_debug(message.header, "entry changed while writing sectors");
                self.write_prepare_release(write, null);
                return;
            }

```

jorangreef replied:
Great, agreed.

jorangreef commented on src/vsr/journal.zig:
```diff
@@ -1696,9 +1729,16 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
+            // Release the write prior to returning control to the caller.
+            // This allows us to enforce journal.writes.len≤1 when replica_count=1, because the
+            // callback may immediately start the next write.
             self.writes.release(write);
```
Is there a scenario where releasing control to another caller ahead of the callback may invalidate other assertions in the callback? For example, what if the other caller changes the same op?

sentientwaffle replied:
> releasing control to another caller ahead of the callback

I'm not sure what you mean here — what other caller are you referring to?

jorangreef replied:
Another concurrent I/O task with it's own caller and callback. Does that make sense?

sentientwaffle replied:
I still don't understand.

```zig
        fn write_prepare_release(self: *Self, write: *Self.Write, wrote: ?*Message) void {
...

            self.writes.release(write); // ← (This used to be at the very end).
            write_callback(replica, wrote, write_trigger);
            replica.message_bus.unref(write_message);
}
```

- The `write_callback()` is being called before the end of `write_prepare_release`, so from the perspective of the journal context which calls `write_prepare_release`, nothing has changed.
- Within `write_callback` doesn't care about whether the write is released or not, except that an additional write might be available to use (this is the goal of the change).

`write_prepare_release` is called at the end of a callback chain to clean up, so I don't think there are any assertions we could collide with. But even if it wasn't, I don't think this change would affect the caller's invariants.

jorangreef replied:
Thanks @sentientwaffle — that's perfect.

jorangreef commented on src/vsr/journal.zig:
```diff
@@ -1837,24 +1855,31 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
         pub fn writing(self: *Self, op: u64, checksum: u128) bool {
+            const slot = self.slot_for_op(op);
             var it = self.writes.iterate();
+            var found: bool = false;
```
```suggestion
            var found: bool = false;
            var it = self.writes.iterate();
```

To keep the `it` variable closer to where it's used in the loop below, and also since `found` is the function's return value and more significant.

jorangreef commented on src/vsr/journal.zig:
```diff
@@ -1837,24 +1855,31 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
                     assert(self.dirty.bit(self.slot_for_op(op)));
-                    return true;
+                    found = true;
+                } else if (write_slot.index == slot.index) {
+                    // If the in-progress write of '{op, checksum}' will be overwritten by another
+                    // write to the same slot, writing() must return false.
+                    found = false;
```
To check if I understand, here we have the op write in progress, and then we see that a different op is going to overwrite the same slot, so that the caller can then overwrite that different op again?

How do we use `writing()` externally?

sentientwaffle replied:
> To check if I understand, here we have the op write in progress, and then we see that a different op is going to overwrite the same slot, so that the caller can then overwrite that different op again?

Correct.

> How do we use `writing()` externally?

Two locations:

1. `Replica.repair_prepare()`: We skip repair a prepare if we are already writing it (either by starting a write if we have it in the pipeline, or sending off a `request_prepare` otherwise).
2. `Replica.write_prepare()`: Don't start writing the prepare if it is already being written.

So if the write queue looks like: `[A, B]` where `A` and `B` are different versions of the same op, and then the replica hits a view change and the new leader picks `A` again, we _should_ queue up a write to `A`, because our pending write to `A` is going to be overwritten by `B`.

I suppose that when the second `A` calls `write_prepare`, we could even scan through the list and remove `B` (which is still waiting for its range lock), and then just allow the first `A` call to do the actual write, though that seems like it would be hard to get right (write?), and difficult to test.

jorangreef replied:
Agreed that we shouldn't try and compress the list. My gut feel too. Thanks for the explanation.

jorangreef reviewed (commented):
Thanks @sentientwaffle — I've finished looking at the WAL. Next is Replica.

jamii commented on src/vsr/superblock.zig:
```diff
@@ -672,46 +718,44 @@ pub fn SuperBlockType(comptime Storage: type) type {
         }
 
         fn write_staging(superblock: *SuperBlock, context: *Context) void {
```
Since `staging` doesn't exist any more, I guess this should be `write_writing`? Unfortunate :D

jamii replied:
Similarly for `const staging: *SuperBlockSector = superblock.writing;` in several places.

sentientwaffle replied:
Yeah, I left it as `write_staging` because even though we don't have a variable called "staging", we are still (conceptually) writing the staged data.

sentientwaffle committed:
WAL: Set prepare_checksums correctly after write

sentientwaffle committed:
SuperBlock: Add Fork and ParentSuperseded errors

sentientwaffle committed:
SuperBlock: Add handling for new quorum errors

And rename "ParentSuperseded" to "ParentSkipped"; the latter is more descriptive.

sentientwaffle committed:
WAL: Clarify condition

It was correct before, but misleading since we are comparing index vs count.

sentientwaffle committed:
VSR: Explain/assert: do not repair old ops

sentientwaffle committed:
VSR/SuperBlock: Rename "writing" to "staging"

`SuperBlock.writing` isn't always being written; `SuperBlock.staging` is a more accurate identifier.

sentientwaffle committed:
CI: zig fmt

sentientwaffle commented:
I think this is ready to merge (pending review approval).

jorangreef reviewed (approved):
I've gone through the latest commits—good assertions and comments.
